### ISSUE TITLE
R Session results in high memory pressure on Mac OS 10.12.3, causes R Studio to abort session

### ISSUE THREAD

**Issue by debshila** [rstudio/rstudio#2340]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.1.419
    OS Version      : Mac OS High Sierra 10.13.3
    Device:              MacBook Pro (Retina, 13-inch, Early 2015)

```
> version
platform       x86_64-apple-darwin15.6.0   
arch           x86_64                      
os             darwin15.6.0                
system         x86_64, darwin15.6.0        
status                                     
major          3                           
minor          4.3                         
year           2017                        
month          11                          
day            30                          
svn rev        73796                       
language       R                           
version.string R version 3.4.3 (2017-11-30)
nickname       Kite-Eating Tree
```
          

### Steps to reproduce the problem
I am unsure of the triggers, but my R session results in high memory pressure resulting in R studio to crash and start a new session. 

[An image showing R Session and R Studio on Activity Monitor](https://drive.google.com/file/d/1frmKKKPNkY5Oyyebkh7mbfepR9iAm-UH/view?usp=sharing)

### Problem Description
This has been happening since I upgraded to High Sierra on my MacBook Pro. The intensity of the computations or size of datasets available in the environment don't seem to matter. Memory shoots up from one or a few GBs to ~42 GB and the R Studio session freezes. The start new session pop up either comes on its own or I have to force quit the R Session in order to get the start new session pop up. This has been a frequent occurrence since my upgrade, sometimes as often as 1 crash an hour.
A coworker also seems to be having the same issue on his newer MacBook Pro.

### Expected Behavior

Memory pressure never resulted in a crash when I was on an older Mac OS platform. This only started after I upgraded to High Sierra. I do not expect my computations to cause my RSession memory pressure to go up to >40 GB. 

1. The output of sessionInfo() 

```
R version 3.4.3 (2017-11-30)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.3

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.4.3 tools_3.4.3    yaml_2.1.16   
```

2. The R code in question : It's not the same piece of code that triggers the crash. I am unsure if any of my code causes R Studio to crash.

3. A diagnostics report: [Here is the diagnostics report after I restarted R Studio. The report was too long, therefore I couldn't directly paste it in.](https://drive.google.com/file/d/1vT2Y5CTM2gn6XhPcEp59B0em2h3Avopv/view?usp=sharing)

**Comment by kevinushey** [comment]
Thanks for the bug report -- unfortunately, nothing stands out from your diagnostics report.

This is a shot in the dark, but we recently had a somewhat similar report from a user who was working with RStudio within a directory containing a very, very large number of files (almost a million files). Is that true in your case?

If you can find anything that seems to be correlated with the hang you're experienced, please let us know and that will give us a better shot at figuring out what's going on here.

**Comment by debshila** [comment]
@kevinushey Thanks for your response. By directory do you mean the immediate parent folder? The directory I am working in does not contain a million files. 
So far, I have not been able to detect what might be correlated with the crash. I will keep you posted if I do. I will ask ask my coworker to be vigilant in case he can figure out what is causing the crash for him.

**Comment by debshila** [comment]
I am not sure if this adds anything to the diagnostics report that I already posted, but I took a sample of the activity on Activity monitor when the memory pressure shot up. Here it is, just in case:
[Activity monitor sample of activity](https://drive.google.com/file/d/1DuDkj7ut_7m9KKcAF261hxP87Za7YcdW/view?usp=sharing)

**Comment by kevinushey** [comment]
Thanks! This is helpful.

Are you working with an R Notebook when this issue occurs? It looks as though RStudio is attempting to render an R Notebook, as per the line:

> + 2234 rstudio::session::modules::rmarkdown::notebook::(anonymous namespace)::createNotebookFromCache(rstudio::core::json::JsonRpcRequest const&, rstudio::core::json::JsonRpcResponse*)  (in rsession) + 590  [0x1004d18ae]

RStudio will generate a preview of your current R Notebook every time you save the document. Perhaps our Notebook save implementation is causing the problem here? You could try testing this with something like:

```R
Rprof("rstudio-trace.Rprof")
.rs.createNotebookFromCache("/path/to/your/notebook.Rmd")
Rprof(NULL)
```

**Comment by kevinushey** [comment]
The trace seems to suggest that most of the time is spent updating the knit metadata, and concatenating R objects:

```R
> summaryRprof("~/Desktop/rstudio-trace.Rprof")
$by.self
                       self.time self.pct total.time total.pct
"c"                        18.94    68.67      18.94     68.67
"knitr::knit_meta_add"      8.26    29.95      27.20     98.62
< ... >
```

Any chance you could try debugging / tracing that `knitr::knit_meta_add()` function to see what's being added / executed there?

As a workaround, I think ensuring that you don't have `html_notebook` as an output type in your YAML header will ensure this background Notebook creation doesn't occur.

**Comment by kevinushey** [comment]
It looks like the `knit_meta_id` attribute may be growing out of control. Unfortunately, I'm not sure why.

What kind of objects are you generating and printing in your R Notebook? Any chance you can provide a reproducible example?

**Comment by debshila** [comment]
@kevinushey These are mostly text, ggplots, summary tables using kable, stargazer tables, mixed effect binomial models, and linear models.

**Comment by kevinushey** [comment]
I see. Unfortunately without a reproducible example I don't think I can help debug further. If you can distill your document down into a minimal case that reproduces the issue you're seeing I can try to look further.

**Comment by debshila** [comment]
I will try to see if I can reproduce the issue with an example, and let you know. Thanks again for all your help!

**Comment by debshila** [comment]
@kevinushey Now I have removed the notebook related YAML from the markdown, I am still being able to use inline outputs. I assumed that inline outputs can only be used in notebook mode. Is that no longer true?

**Comment by kevinushey** [comment]
Inline outputs can actually be used in plain R Markdown documents (ie, when the target output is not an R Notebook). The only difference is that RStudio doesn't attempt to automatically generate a Notebook for you behind the scenes; the final document must instead be generated through a full `rmarkdown::render()` call (or similar).

**Comment by nikkoc** [comment]
I know I'm a bit late to the game here, but I recently ran into a similar issue where I found that the `knit_meta` was not being appropriately cleaned as expected. The following is the same comment as I gave on issue #2319. I know the situation is a bit different, but it looks like they could both come down to `knitr::knit_meta()`.

I've been having an issue that sounds just like this. My Shiny app generates a PDF report from an Rmd file as part of a download handler. The PDF that I test with is about 7 pages. By testing the same download button in the same session, the first 4-5 PDF renders go smoothly. However, after the nth iteration, the rsession memory jumps sizably (to ~ 1GB, up from ~250MB average), then n+1 iteration the memory jumps into the 7-8 GB range and climbs.

After stepping the code inside `render`, I found that [old_knit_meta <- knit_meta_reset()](https://github.com/rstudio/rmarkdown/blob/bbd0786d82f82d9eaef7f0b9e2c553b523407161/R/render.R#L417) was somehow not getting cleared. For example, whereas it should typically register around 40 bytes for an empty list, it would register 843317640 bytes. `knit_meta_reset()` calls `knitr::knit_meta()`, which defaults to `clean=T`, so it should clear the `.knitEnv$meta` after it returns it. It usually does, but for some weird reason that I don't know, it doesn't in this circumstance and thus causes old_knit_meta to get really large.

What I found was that if I 'reset' the knit_meta myself by calling `knitr::knit_meta(clean=T)` when debugging the code in the ~1GB memory iteration, it would release that memory and go back to the ~250MB range (following @areedv 's comment in issue #2319 , I call `gc()` after `render`). As a result I now 'force' knit_meta to be empty by calling `knit_meta` before I call `rmarkdown::render`. I made a function specifically to wrap the `knit_meta` call inside `invisible()` to avoid printing the results.

I have only done this today, so I haven't run into the issue YET. But when I tested this solution, I grew tired of running the same PDF after about 25 times and I did not see any anomalous spike in memory consistent with this bug.

**Comment by kevinushey** [comment]
Okay, I can finally reproduce this issue. Here's an example. Create an R Notebook with the contents:

`````
---
title: "Untitled"
output: html_notebook
---

```{r}
library(kableExtra)
kbl <- kableExtra::kable(mtcars, format = "latex")
kableExtra::save_kable(kbl, tempfile())
```
`````

Then, try to run that chunk. When I do so I get an error (it may or may not be related):

<img width="751" alt="screen shot 2019-02-27 at 10 04 19 pm" src="https://user-images.githubusercontent.com/1976582/53544643-af037900-3adb-11e9-8e03-3ad0b607bce6.png">

I also see that `.knitEnv$meta` is populated:

```
> str(knitr:::.knitEnv$meta)
List of 14
< ... latex dependencies omitted ... >
 - attr(*, "knit_meta_id")= chr [1:196] "" "" "" "" ...
```

Now, try editing and saving the document. This will force a render behind the scenes. When I do this, I then see:

```
> str(knitr:::.knitEnv$meta)
List of 14
< ... latex dependencies omitted ... >
 - attr(*, "knit_meta_id")= chr [1:2744] "" "" "" "" ...
```

In other words, the size of `knit_meta_id` grows multiplicatively with each subsequent render.

This is the definition of `knit_meta_add()`:

```
function (meta, label = "") 
{
  if (length(meta)) {
    meta_id = attr(.knitEnv$meta, "knit_meta_id")
    .knitEnv$meta = c(.knitEnv$meta, meta)
    attr(.knitEnv$meta, "knit_meta_id") = c(meta_id, rep(label, 
      length(meta)))
  }
  .knitEnv$meta
}
```

And `rmarkdown` is calling it as:

```
knitr::knit_meta_add(old_knit_meta, attr(old_knit_meta, 'knit_meta_id'))
```

so effectively, we end up replicating the knit_meta_id (as the label) far more times than we should.

Ultimately this is a bug to be fixed on the `rmarkdown` side so I'm going to file this there.

**Comment by kevinushey** [comment]
Reproduced independently of RStudio and filed here: https://github.com/rstudio/rmarkdown/issues/1538

**Comment by kevinushey** [comment]
Fixed in the development version of knitr.
